arangod/Transaction/Methods.cpp

```diff
@@ -409,1386 +409,1607 @@ OperationResult emptyResult(OperationOptions const& options) {
-transaction::Methods::~Methods() {
-  if (_mainTransaction) {  // _nestingLevel == 0
-    // unregister transaction from context
-    _transactionContext->unregisterTransaction();
+  Future<OperationResult> execute() {
+    _replicationData->openArray(true);
+    std::unordered_map<ErrorCode, size_t> errorCounter;
+    Result res;
+    if (this->_value.isArray()) {
+      this->_resultBuilder.openArray();
 
-    // auto abort a non-read-only and still running transaction
-    if (_state->status() == transaction::Status::RUNNING) {
-      if (_state->isReadOnlyTransaction()) {
-        // read-only transactions are never comitted or aborted during their
-        // regular life cycle. we want now to properly clean up and count them.
-        _state->updateStatus(transaction::Status::COMMITTED);
-      } else {
-        auto res = this->abort();
+      for (VPackSlice s : VPackArrayIterator(this->_value)) {
+        res = this->self().processValue(s, true);
         if (res.fail()) {
-          LOG_TOPIC("6d20f", ERR, Logger::TRANSACTIONS)
-              << "Abort failed while destroying transaction " << tid()
-              << " on server " << ServerState::instance()->getId() << " "
-              << res;
+          createBabiesError(
+              _replicationType == Methods::ReplicationType::FOLLOWER
+                  ? nullptr
+                  : &this->_resultBuilder,
+              errorCounter, res);
+          res.reset();
         }
       }
+
+      this->_resultBuilder.close();
+    } else {
+      res = this->self().processValue(this->_value, false);
+
+      // on a follower, our result should always be an empty object
+      if (_replicationType == Methods::ReplicationType::FOLLOWER) {
+        TRI_ASSERT(this->_resultBuilder.slice().isNone());
+        // add an empty object here so that when sending things back in JSON
+        // format, there is no "non-representable type 'none'" issue.
+        this->_resultBuilder.add(VPackSlice::emptyObjectSlice());
+      }
     }
+    _replicationData->close();
+
+    // on a follower, our result should always be an empty array or object
+    TRI_ASSERT(_replicationType != Methods::ReplicationType::FOLLOWER ||
+               (this->_value.isArray() &&
+                this->_resultBuilder.slice().isEmptyArray()) ||
+               (this->_value.isObject() &&
+                this->_resultBuilder.slice().isEmptyObject()));
+    TRI_ASSERT(_replicationData->slice().isArray());
+    TRI_ASSERT(_replicationType != Methods::ReplicationType::FOLLOWER ||
+               _replicationData->slice().isEmptyArray());
+    TRI_ASSERT(!this->_value.isArray() || this->_options.silent ||
+               this->_resultBuilder.slice().length() == this->_value.length())
+        << "operation: " << _operationType
+        << ", silent: " << this->_options.silent
+        << ", replicationType: " << static_cast<int>(this->_replicationType)
+        << ", isArray: " << this->_value.isArray();
```

Comment:
This should be removed, because if it triggers, it may crash the AssertionLogger with stack overflow!

Reply:
I have adapted the output so that we no longer print the actual payload, so I think this should be fine.

Reply:
Ok, we can leave it here, although I don't think the message will be great help now. Anyway, it's fine.